src/structs/lepton_decoder.rs

```diff
@@ -121,105 +121,40 @@ fn decode_row_wrapper<R: Read>(
     curr_y: i32,
     features: &EnabledFeatures,
 ) -> Result<()> {
-    let mut context = image_data.off_y(curr_y);
+    let mut block_context = image_data.off_y(curr_y);
+
+    let left_model;
+    let middle_model;
 
-    let block_width = image_data.get_block_width();
     if is_top_row[component] {
         is_top_row[component] = false;
-        decode_row(
-            model,
-            bool_reader,
-            &qt,
-            &pts.corner[component],
-            &pts.top[component],
-            &pts.top[component],
-            image_data,
-            &mut context,
-            neighbor_summary_cache,
-            component_size_in_blocks[component],
-            features,
-        )
-        .context(here!())?;
-    } else if block_width > 1 {
-        let _bt = component;
-        decode_row(
-            model,
-            bool_reader,
-            &qt,
-            &pts.mid_left[component],
-            &pts.middle[component],
-            &pts.mid_right[component],
-            image_data,
-            &mut context,
-            neighbor_summary_cache,
-            component_size_in_blocks[component],
-            features,
-        )
-        .context(here!())?;
+
+        left_model = &pts.corner[component];
+        middle_model = &pts.top[component];
     } else {
-        assert!(block_width == 1, "block_width == 1");
-        decode_row(
-            model,
-            bool_reader,
-            &qt,
-            &pts.width_one[component],
-            &pts.width_one[component],
-            &pts.width_one[component],
-            image_data,
-            &mut context,
-            neighbor_summary_cache,
-            component_size_in_blocks[component],
-            features,
-        )
-        .context(here!())?;
+        left_model = &pts.mid_left[component];
+        middle_model = &pts.middle[component];
     }
 
-    Ok(())
-}
-
-fn decode_row<R: Read>(
-    model: &mut Model,
-    bool_reader: &mut VPXBoolReader<R>,
-    qt: &QuantizationTables,
-    left_model: &ProbabilityTables,
-    middle_model: &ProbabilityTables,
-    right_model: &ProbabilityTables,
-    image_data: &mut BlockBasedImage,
-    block_context: &mut BlockContext,
-    neighbor_summary_cache: &mut [NeighborSummary],
-    component_size_in_blocks: i32,
-    features: &EnabledFeatures,
-) -> Result<()> {
+    let component_size_in_blocks = component_size_in_blocks[component];
     let block_width = image_data.get_block_width();
-    if block_width > 0 {
-        parse_token::<R, false>(
-            model,
-            bool_reader,
-            image_data,
-            block_context,
-            neighbor_summary_cache,
-            qt,
-            left_model,
-            features,
-        )
-        .context(here!())?;
-        let offset = block_context.next();
 
-        if offset >= component_size_in_blocks {
-            return Ok(()); // no sure if this is an error
-        }
-    }
+    for jpeg_x in 0..block_width {
```

Comment:
Heh, this can be simplified to `loop` - checking offset against min(beginning of next line, component_size_in_blocks).

Reply:
Have no permission to push here, realized [this](https://github.com/Melirius/lepton_jpeg_rust/commit/57dbb04254978bbe08a421bbbd3aa5a2f2ebdcde)

Reply:
Do you think it will be better?